solutions/gold/rmi-17-Hangman2.mdx

```diff
@@ -0,0 +1,154 @@
+---
+id: rmi-17-Hangman2
+source: RMI 2017 Day 1
+title: Hangman 2
+author: Justin Ji
+---
+
+<Spoiler title = "Hint">
+
+Break the problem into two cases: $n < k$ and $n \geq k$.
+
+</Spoiler>
+
+<Spoiler title = "Hint 2">
+
+Consider a $\mathcal{O}(NKmin(N, K))$ solution. What is $min(n, k)$ bounded by?
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+First, note that $min(n, k) \leq \sqrt{nk}$. This will be important for
+analyzing the time complexity of our solution. Now, let's consider the two
+cases given in the hint.
+
+### Case 1: $n < k$
+Because $n < \sqrt{nk}$, we can brute force each pair of strings and check if
+they differ by no more than two characters. This takes $\mathcal{O}(N^2K)$ time,
+which is equivalent to $\mathcal{O}(NK\sqrt{NK})$ time.
+
+### Case 2: $n \geq k$
+Let's compare two strings. Note how allowing at most two characters to be
+different is equivalent to ignoring two characters in both strings and checking
+for equality. Thus, we can brute force on every pair of characters that we are
+ignoring, and then check for equality between strings.
+
+Let the pair of characters we are brute forcing be $(i, j)$. Then,
+we have the following intervals that need to be identical in both strings:
+
+$[0, i - 1], [i + 1, j - 1], [j + 1, k - 1]$
+
+Now, if we want to compare our two strings, then we just need the substrings to
+be equal. This can be checked using string hashing.
+
+For each string in our list, we need to find another string that has the same
+hashes. This is best done by sorting all the tuples of hashes, and then checking
+adjacent items and seeing if the hashes are equal.
+
+Overall, this takes $\mathcal{O}(NK^2\log({N}))$ time. With a bit more analysis,
+you can reach the time complexity of $\mathcal{O}(NK\sqrt{NK}\log(\sqrt{NK}))$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NK\sqrt{NK}\log(\sqrt{NK}))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// BeginCodeSnip{String Hashing Template}
+class HashedString {
+  private:
+	// change M and B if you want
+	static const long long M = 1e9 + 9;
+	static const long long B = 9973;
+
+	// pow[i] contains B^i % M
+	static vector<long long> pow;
+
+	// p_hash[i] is the hash of the first i characters of the given string
+	vector<long long> p_hash;
+
+  public:
+	HashedString() {}
```

Comment:
was this part of the thing

Reply:
nah, I had to add it b/c C++ was complaining about `vector<HashedString>`

Reply:
i mean hashedstring expects a string,,, why do you need them to start out empty

Reply:
> i mean hashedstring expects a string,,, why do you need them to start out empty

fair enough, should i just put an empty string as the fill value then